fix: base_url_fix removing URI from baseURL

abhishekgarg18 opened the pull request:
Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes
- [ ] If data sources for any opportunity has been updated/added, please update the [wiki](https://wiki.corp.adobe.com/display/AEMSites/Data+Sources+for+Opportunities) for same opportunity.

## Related Issues


Thanks for contributing!

## Issue
When processing meta tag suggestions for URLs like `https://kpmg.com/ch/en.html`, the system is incorrectly appending path segments, resulting in malformed URLs:

- Current: `https://kpmg.com/ch/en.html/ch/en/about.html`
- Expected: `https://kpmg.com/ch/en/about.html`

The issue occurs because we're directly concatenating the endpoint to the base URL without proper path handling, causing path segments to be duplicated.

## Solution
Added `useHostnameOnly` configuration in site's delivery config to control URL handling:
- When `true`: Strips URL to just protocol + hostname (e.g., `https://kpmg.com`)
- When `false` (default): Preserves the full path, only removing trailing slashes

This allows sites to configure how their base URLs should be handled, which is particularly important for sites with complex URL structures like KPMG where we need to prevent path segment duplication.

## Implementation
- Added `useHostnameOnly` flag in delivery config
- Updated `getBaseUrl` to respect this configuration
- Added proper error handling and logging
- Default behavior remains unchanged (false) for backward compatibility

## Testing
Added test cases for:
- URLs with port numbers
- URLs with query parameters
- Error handling scenarios
- Default behavior verification

absarasw commented on src/metatags/opportunity-utils.js:
```diff
@@ -11,7 +11,13 @@
  */
 
 export function removeTrailingSlash(url) {
```
Can we create a new function as getBaseUrl and use that? This might be used somewhere where we expect to only to remove the trailing slash.

abhishekgarg18 commented:
@absarasw @dipratap please review.

dipratap commented:
@abhishekgarg18 can you please add a brief PR description of the change?

dipratap commented on src/metatags/handler.js:
```diff
@@ -47,6 +47,9 @@ export async function opportunityAndSuggestions(finalUrl, auditData, context) {
   const { log } = context;
   const { detectedTags } = auditData.auditResult;
+  const site = opportunity.getSite?.();
+  const deliveryConfig = site !== null ? site?.getDeliveryConfig() : null;
```
Can you please share the purpose of delivery config?

abhishekgarg18 replied:
I tried using config, but I am not able to add new value in this "config", so reverted to use deliveryConfig
Cc : @absarasw @dipratap

dipratap commented on src/metatags/handler.js:
```diff
@@ -47,6 +47,9 @@ export async function opportunityAndSuggestions(finalUrl, auditData, context) {
   const { log } = context;
   const { detectedTags } = auditData.auditResult;
+  const site = opportunity.getSite?.();
+  const deliveryConfig = site !== null ? site?.getDeliveryConfig() : null;
```
```suggestion
const useHostnameOnly = opportunity.getSite?.()
  ?.getDeliveryConfig()
  ?.useHostnameOnly ?? false;
```

dipratap commented on src/metatags/opportunity-utils.js:
```diff
@@ -14,6 +14,19 @@ export function removeTrailingSlash(url) {
   return url.endsWith('/') ? url.slice(0, -1) : url;
 }
 
+export function getBaseUrl(url, useHostnameOnly = true) {
```
useHostnameOnly default should be false. If the urls structure change then all the previous suggestions, detected before this change, will have different urls than the new ones - which will lead to duplicate suggestions

dipratap replied:
Before that, I would like to understand why are we doing this change, as it can break things.

abhishekgarg18 replied:
changed default value to `false`

dipratap commented:
There is a overrideBaseUrl param that is currently used to overwrite the site baseurl params, we should check if the final Url param, which we are not using in this case, uses that overrideBaseUrl config, and if yes what should be the behaviour.

abhishekgarg18 commented:
@dipratap I have added a description of PR.

Currently, we are using finalUrl,
``` url: getBaseUrl(auditData.auditResult.finalUrl, useHostnameOnly) + endpoint,```
i added 
```"fetchConfig": {
            "overrideBaseURL": "https://kpmg.com"
        }```
        
  it still didn't work.

solaris007 commented:
:tada: This PR is included in version 1.157.8 :tada:

The release is available on [GitHub release](https://github.com/adobe/spacecat-audit-worker/releases/tag/v1.157.8)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
